src/eth_provider/database/ethereum.rs

```diff
@@ -199,3 +199,189 @@ impl EthereumBlockStore for Database {
+        );
+    }
+
+    async fn test_upsert_pending_transactions(unstructured: &mut arbitrary::Unstructured<'_>, database: &Database) {
+        // Generate 10 pending transactions and add them to the database
+        let pending_transactions: Vec<StoredPendingTransaction> =
+            (0..10).map(|_| StoredPendingTransaction::arbitrary_with_optional_fields(unstructured).unwrap()).collect();
+
+        // Add pending transactions to the database
+        for tx in &pending_transactions {
+            database
+                .upsert_pending_transaction(tx.tx.clone(), tx.retries)
+                .await
+                .expect("Failed to update pending transaction in database");
+        }
+
+        // Test retrieving a pending transaction by its hash
+        let first_pending_transaction = pending_transactions.first().unwrap();
+        assert_eq!(
+            database.pending_transaction(&first_pending_transaction.tx.hash).await.unwrap(),
+            Some(first_pending_transaction.tx.clone())
+        );
+
+        // Test retrieving a non-existent pending transaction by its hash
+        let unstored_transaction = StoredTransaction::arbitrary_with_optional_fields(unstructured).unwrap().tx;
+        assert_eq!(database.pending_transaction(&unstored_transaction.hash).await.unwrap(), None);
+
+        // Test retrieving the number of retries for a pending transaction
+        assert_eq!(
+            database.pending_transaction_retries(&first_pending_transaction.tx.hash).await.unwrap(),
+            first_pending_transaction.clone().retries.saturating_add(1)
+        );
+
+        // Test retrieving the number of retries for a non-existent pending transaction
+        assert_eq!(database.pending_transaction_retries(&unstored_transaction.hash).await.unwrap(), 0);
+    }
+
+    async fn test_upsert_transactions(unstructured: &mut arbitrary::Unstructured<'_>, database: &Database) {
+        // Generate and upsert a mock transaction into the database
+        let mock_transaction = StoredTransaction::arbitrary_with_optional_fields(unstructured).unwrap();
+        database.upsert_transaction(mock_transaction.clone().tx).await.unwrap();
+
+        // Test retrieving an upserted transaction by its hash
+        assert_eq!(database.transaction(&mock_transaction.tx.hash).await.unwrap(), Some(mock_transaction.tx.clone()));
+
+        // Generate and upsert a mock pending transaction into the database
+        let mock_pending_transaction = StoredPendingTransaction::arbitrary_with_optional_fields(unstructured).unwrap();
+        database
+            .upsert_pending_transaction(mock_pending_transaction.clone().tx, mock_pending_transaction.clone().retries)
+            .await
+            .unwrap();
+
+        // Test retrieving an upserted pending transaction by its hash
+        assert_eq!(
+            database.pending_transaction(&mock_pending_transaction.clone().tx.hash).await.unwrap(),
+            Some(mock_pending_transaction.clone().tx)
```

Comment:
I don't think you need either `clone` here

Reply:
True, just fixed